File: ios/MenCare/Core/Model/Crm.swift
import Foundation

enum ClientSegment: CaseIterable, Hashable {
    case tutti, fedeli, inattivi60, noShow, topSpesa
}

struct ClientRecord: Identifiable, Hashable {
    let id: String
    var firstName: String
    var lastName: String
    var phone: String
    var email: String
    let customerSince: LocalDate
    var visitCount: Int
    var lifetimeSpendCents: Int64
    var noShowCount: Int
    var lastVisit: LocalDate?
    var preferredServiceIds: [String] = []
    var preferredOperatorId: String?
    var marketingOptIn: Bool = true
    /// Habits from the client sheet (`GET /crm/clients/:id`): nil in lists.
    var favoriteOperatorId: String?
    var averageDaysBetweenVisits: Int?

    var fullName: String { "\(firstName) \(lastName)" }
    var initials: String {
        let f = firstName.first.map(String.init) ?? ""
        let l = lastName.first.map(String.init) ?? ""
        return (f + l).trimmingCharacters(in: .whitespaces)
    }

    func isInactiveSince(_ today: LocalDate, days: Int = 60) -> Bool {
        guard let lastVisit else { return true }
        return lastVisit.plusDays(days) < today
    }

    func isLoyal(_ today: LocalDate) -> Bool {
        visitCount >= 10 && !isInactiveSince(today)
    }
}

/// Time window the owner dashboard aggregates over.
enum DashboardPeriod: CaseIterable, Hashable {
    case day, week, month
}

struct DashboardStats: Hashable {
    var period: DashboardPeriod
    var revenueCents: Int64
    var revenueTrendPercent: Int
    /// Non-cancelled appointments of every operator in the period.
    var appointmentCount: Int
    var noShowPercent: Double
    var averageTicketCents: Int64
    var operatorOccupancy: [OperatorOccupancy]
    /// The next 7 days from today, whatever the period: where there is room to fill.
    var upcomingDays: [UpcomingDay]
    var inactiveClients: [ClientRecord]
}

struct OperatorOccupancy: Hashable, Identifiable {
    let operatorId: String
    let operatorName: String
    let percent: Int

    var id: String { operatorId }
}

/// One of the next 7 days on the dashboard: how full the salon is, how many
/// clients asked "Avvisami" for it, and whether it is closed.
struct UpcomingDay: Hashable, Identifiable {
    let date: LocalDate
    let occupancyPercent: Int
    let waitlistCount: Int
    let closed: Bool

    var id: LocalDate { date }
}

File: android/core/model/src/main/kotlin/com/devora/mencare/core/model/Crm.kt
package com.devora.mencare.core.model

import java.time.LocalDate

enum class ClientSegment { TUTTI, FEDELI, INATTIVI_60, NO_SHOW, TOP_SPESA }

data class ClientRecord(
    val id: String,
    val firstName: String,
    val lastName: String,
    val phone: String,
    val email: String,
    val customerSince: LocalDate,
    val visitCount: Int,
    val lifetimeSpendCents: Long,
    val noShowCount: Int,
    val lastVisit: LocalDate?,
    val preferredServiceIds: List<String> = emptyList(),
    val preferredOperatorId: String? = null,
    val marketingOptIn: Boolean = true,
    /** Habits from the client sheet (`GET /crm/clients/:id`): null in lists. */
    val favoriteOperatorId: String? = null,
    val averageDaysBetweenVisits: Int? = null,
) {
    val fullName: String get() = "$firstName $lastName"
    val initials: String get() = "${firstName.firstOrNull() ?: ' '}${lastName.firstOrNull() ?: ' '}".trim()

    fun isInactiveSince(today: LocalDate, days: Long = 60): Boolean =
        lastVisit == null || lastVisit.plusDays(days) < today

    fun isLoyal(today: LocalDate): Boolean =
        visitCount >= 10 && !isInactiveSince(today)
}

/** Time window the owner dashboard aggregates over. */
enum class DashboardPeriod { DAY, WEEK, MONTH }

data class DashboardStats(
    val period: DashboardPeriod,
    val revenueCents: Long,
    val revenueTrendPercent: Int,
    /** Non-cancelled appointments of every operator in the period. */
    val appointmentCount: Int,
    val noShowPercent: Double,
    val averageTicketCents: Long,
    val operatorOccupancy: List<OperatorOccupancy>,
    /** The next 7 days from today, whatever the period: where there is room to fill. */
    val upcomingDays: List<UpcomingDay>,
    val inactiveClients: List<ClientRecord>,
)

data class OperatorOccupancy(val operatorId: String, val operatorName: String, val percent: Int)

/**
 * One of the next 7 days on the dashboard: how full the salon is, how many
 * clients asked "Avvisami" for it, and whether it is closed.
 */
data class UpcomingDay(
    val date: LocalDate,
    val occupancyPercent: Int,
    val waitlistCount: Int,
    val closed: Boolean,
)
